### PR TITLE
rgw: S3 bucket logging

### PR DESCRIPTION
**Pull request by caisan**
This PR is about the implementation of S3 bucket logging for RGW.
More details :http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTlogging.html
#19689

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4425,7 +4425,84 @@ void RGWDeleteLC::execute()
...
+  oid = bl_oid_prefix;
+  char buf[32];
+  snprintf(buf, 32, ".%d", index);
+  oid.append(buf);
+  return;
+}
+void RGWPutBL::execute()
+{
+  bufferlist bl;
+  RGWBucketloggingConfiguration_S3 *config = NULL;
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: nullptr

File: src/rgw/rgw_bl.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+};
+class RGWBucketloggingConfiguration
+{
+protected:
+    CephContext *cct;
+    map<string, string> blrule;
+public:
+    RGWBucketloggingConfiguration(CephContext *_cct) : cct(_cct) {}
+    RGWBucketloggingConfiguration() : cct(NULL) {}
```
Comment (mattbenjamin) [inline src/rgw/rgw_bl.h]: nullptr

File: src/rgw/rgw_bl.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+#define HASH_PRIME_BL 12289
+#define MAX_ID_LEN 255
+
+static string bl_oid_prefix = "bl";
+
+class BLRule
+{
+protected:
+    string status;
```
Comment (mattbenjamin) [inline src/rgw/rgw_bl.h]: this is expensive, for something that could be bool?

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -3089,10 +3089,11 @@ RGWOp *RGWHandler_REST_Bucket_S3::op_head()
 
 RGWOp *RGWHandler_REST_Bucket_S3::op_put()
 {
-  if (s->info.args.sub_resource_exists("logging"))
-    return NULL;
+  //if (s->info.args.sub_resource_exists("logging"))
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.cc]: introduces dead code

File: src/rgw/rgw_bl.cc [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+#include "rgw_bl.h"
+
+using namespace librados;
+
+static string bl_index_lock_name = "bl_process";
+
+void RGWBucketloggingConfiguration::add_rule(BLRule *rule)
+{
+    if (rule != NULL) {
+        blrule[string("Status")] = string("Enabled");
```
Comment (mattbenjamin) [inline src/rgw/rgw_bl.cc]: noted elsewhere, status

File: src/rgw/rgw_bl_s3.cc [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+{
+    //fixme: rebuild the config
+    int ret =0;    
+    dest = *this;
+    return ret;
+}
+XMLObj* RGWBLXMLParser_S3::alloc_obj(const char* el)
+{
+    XMLObj *obj = NULL;
+    if (strcmp(el, "BucketLoggingStatus") == 0) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_bl_s3.cc]: I'm not very familiar with existing idioms in this part of our code, but looking forward, are there options for reducing heap allocation that we could try here?

File: src/rgw/rgw_bl.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    string& get_targetbucket() {
+        return targetbucket;
+    }
+};
+class RGWBucketloggingConfiguration
+{
+protected:
+    CephContext *cct;
+    map<string, string> blrule;
```
Comment (mattbenjamin) [inline src/rgw/rgw_bl.h]: sorry if off base, but is this an efficient representation?

**Comment by mikulely** [comment]
@caisan  @mattbenjamin  @cbodley It seems that this pr is trying to impl HTTP API of Put Bucket Logging, which has the same functionality as comit https://github.com/ceph/ceph/pull/14841/commits/bac1de68811e7c9ca567c6ca7333b4b459c93f23 . We've fired pr https://github.com/ceph/ceph/pull/14841 to bring s3 bucket logging. Welcome to join the discussion.

File: src/rgw/rgw_rest_s3.cc (hunk 4 above)
Comment (amitkumar50) [inline src/rgw/rgw_rest_s3.cc]: @caisan Even commented code should be removed

**Comment by amitkumar50** [comment]
@caisan PR is not signed.
Please look https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst

**Pull request closed without merging**